README.md
# nv_updater

An update checker plugin for [noveltree](https://peter88213.github.io/noveltree)

For more information, see the [project homepage](https://peter88213.github.io/nv_updater) with description and download instructions.

## Feedback? Ideas? Feature requests?

You can go to the [noveltree "discussions" forum](https://github.com/peter88213/noveltree/discussions) and start a thread.

# Contributing

## How to provide translations

First, you need to know your language code according to ISO 639-1.

For English, this is, for example, `en`, for German, it is `de`.

**NOTE:** The procedure described below is greatly simplified if you create a language pack based on the [noveltree_xx](https://github.com/peter88213/noveltree_xx) template and use the tools provided for this purpose. 

### Create a message catalog

A "message catalog" is a dictionary for the plugin's messages and menu entries.

For creating a message catalog, you download a template with all English messages from [here](https://github.com/peter88213/nv_updater/blob/main/i18n/messages.pot). 


Rename `messages.pot` to `<your language code>.po`, then give some specific information in the header data by modifying the following lines:

```
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language: LANGUAGE\n"
```

**NOTE:** Be sure to use a text editor that writes utf-8 encoded text. Otherwise, it may not work with non-ASCII characters used in your language.

The  `<your language code>.po` dictionary is organized as a set of *message ID (msgid)* - *message string (msgstr)* pairs, where *msgid* means the English term, and *msgstr* means the translated term. This is an example for such a pair where the message string is still missing:

```
msgid "Cannot overwrite file"
msgstr ""
```

Now you enter all missing message strings. 
- If a message ID contains placeholders like `{}`, be sure to put them also into the message string.  
- If a message ID starts with `!`, the message string must also start with `!`. 

Before you distribute your translations, you can convert and install the message catalog for testing. 

### Convert the message catalog to binary format

The plugin needs the message catalog in binary format. This is easily achieved using the **msgfmt.py** converter script. 
You find it in your Python installation, in the **Tools/i18n** subdirectory. If not, you can download the code from [here](https://github.com/python/cpython/blob/main/Tools/i18n/msgfmt.py)

Name the binary file **nv_updater.mo**. 


### Install your translation for testing

Add a subdirectory tree to **noveltree/locale**, and place *nv_updater.mo* there, like this:

```
<your home directory>
└── .novxlib/
    └── noveltree/
        └── locale/
            └─ <language code>/
               └─ LC_MESSAGES/
                  └─ nv_updater.mo
```

Then start *noveltree* and see whether your translation works. 

**NOTE:** At startup, *noveltree* tries to load a message dictionary that fits to the system language. If it doesn't find a matching language code in the *locale* directory, it uses English as default language. 

**HINT:** *nv_updater* comes with German translations. Look at the `de` directory tree, if you need an example. 


### Contribute your translations

If *noveltree* works fine with your translations, you can consider contributing it. 

An easy way may be to put a posting in the [noveltree forum](https://github.com/peter88213/noveltree/discussions), appending your  `<your language code>.po` file. 


## Development

### Conventions

See https://github.com/peter88213/novxlib/blob/main/docs/conventions.md

## Development tools

- [Python](https://python.org) version 3.10.
- [Eclipse IDE](https://eclipse.org) with [PyDev](https://pydev.org) and *EGit*.
- *Apache Ant* is used for building the application.

## License

This is Open Source software, and the *nv_updater* plugin is licensed under GPLv3. See the
[GNU General Public License website](https://www.gnu.org/licenses/gpl-3.0.en.html) for more
details, or consult the [LICENSE](https://github.com/peter88213/nv_updater/blob/main/LICENSE) file.


docs/changelog.md
[Project home page](index) > Changelog

------------------------------------------------------------------------

## Changelog


### v0.2.3

- Refactor: use configparser.

Compatibility: noveltree v4.36 API

### v0.2.2

- Modify message (#1)
- Refactor.
- Provide instructions for use.
- Update documentation.

Compatibility: noveltree v4.36 API

### v0.2.1

- Add debugging code.

Compatibility: noveltree v4.36 API

### v0.2.0

- Check noveltree and installed plugins for updates.
- Allow the user to cancel the update check.

Compatibility: noveltree v4.36 API

### v0.1.0

- Check noveltree for updates.

Compatibility: noveltree v4.36 API


tools/make_pot.py
"""Generate a template file (pot) for message translation.

For further information see https://github.com/peter88213/nv_updater
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import sys
import build_nv_updater as build_plugin
sys.path.insert(0, f'{os.getcwd()}/../../novxlib-Alpha/src')
import pgettext

APP = 'nv_updater'
POT_FILE = '../i18n/messages.pot'


def make_pot(version='unknown'):
    # Generate a complete plugin.
    build_plugin.main()

    # Generate a pot file from the script.
    if os.path.isfile(POT_FILE):
        os.replace(POT_FILE, f'{POT_FILE}.bak')
        backedUp = True
    else:
        backedUp = False
    try:
        pot = pgettext.PotFile(POT_FILE, app=APP, appVersion=version)
        pot.scan_file(build_plugin.TARGET_FILE)
        print(f'Writing "{pot.filePath}"...\n')
        pot.write_pot()

    except:
        if backedUp:
            os.replace(f'{POT_FILE}.bak', POT_FILE)
        print('WARNING: Cannot write pot file.')
        return False

    return True


if __name__ == '__main__':
    try:
        success = make_pot(sys.argv[1])
    except:
        success = make_pot()
    if not success:
        sys.exit(1)


tools/translate_de.py
"""Generate German translation files for GNU gettext.

- Update the project's 'de.po' translation file.
- Generate the language specific 'novxlib.mo' dictionary.

Usage: 
translate_de.py

File structure:

├── novxlib/
│   ├── i18n/
│   │   └── de.json
│   └── src/
│       ├── translations.py
│       └── msgfmt.py
└── nv_updater/
    ├── src/ 
    ├── tools/ 
    │   └── translate_de.py
    └── i18n/
        ├── messages.pot
        ├── de.po
        └── locale/
            └─ de/
               └─ LC_MESSAGES/
                  └─ nv_updater.mo
    
Copyright (c) 2024 Peter Triesberger
For further information see https://github.com/peter88213/nv_updater
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import sys
sys.path.insert(0, f'{os.getcwd()}/../../novxlib-Alpha/src')
import translations
from shutil import copyfile
import msgfmt

APP_NAME = 'nv_updater'
PO_PATH = '../i18n/de.po'
MO_PATH = f'../i18n/locale/de/LC_MESSAGES/{APP_NAME}.mo'
MO_COPY = f'../../kalliope/src/locale/de/LC_MESSAGES/{APP_NAME}.mo'


def main(version='unknown'):
    if translations.main('de', app=APP_NAME, appVersion=version, json=True):
        print(f'Writing "{MO_PATH}" ...')
        msgfmt.make(PO_PATH, MO_PATH)
        copyfile(MO_PATH, MO_COPY)
    else:
        sys.exit(1)


if __name__ == '__main__':
    try:
        main(sys.argv[1])
    except:
        main()


src/novelyst_plugin.py
"""Plugin template for novelyst.

Requires Python 3.6+
Copyright (c) 2023 Peter Triesberger
For further information see https://github.com/peter88213/novelyst_plugin
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import sys
import os
import tkinter as tk
import locale
import gettext
import webbrowser
from novelystlib.plugin.plugin_base import PluginBase

# Initialize localization.
LOCALE_PATH = f'{os.path.dirname(sys.argv[0])}/locale/'
try:
    CURRENT_LANGUAGE = locale.getlocale()[0][:2]
except:
    # Fallback for old Windows versions.
    CURRENT_LANGUAGE = locale.getdefaultlocale()[0][:2]
try:
    t = gettext.translation('novelyst_plugin', LOCALE_PATH, languages=[CURRENT_LANGUAGE])
    _ = t.gettext
except:

    def _(message):
        return message


class Plugin(PluginBase):
    """Template plugin class.
    
    Public methods:
        install(ui) -- Install the plugin and extend the novelyst user interface.
    """
    VERSION = '@release'
    NOVELYST_API = '4.30'
    DESCRIPTION = 'Plugin template'
    URL = 'https://peter88213.github.io/novelyst_plugin'
    _HELP_URL = 'https://peter88213.github.io/novelyst_plugin/usage'

    def install(self, ui):
        """Install the plugin and extend the novelyst user interface.
        
        Positional arguments:
            ui -- reference to the NovelystTk instance of the application.
        """
        self._ui = ui

        # Add an entry to the Help menu.
        self._ui.helpMenu.add_command(label=_('novelyst_plugin Online help'), command=lambda: webbrowser.open(self._HELP_URL))



src/novelyst_updater.py
"""An update checker plugin for novelyst.

Requires Python 3.6+
Copyright (c) 2023 Peter Triesberger
For further information see https://github.com/peter88213/novelyst_updater
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

"""
import sys
import os
from tkinter import messagebox
import locale
import gettext
from urllib.request import urlopen
import webbrowser
import configparser

# Initialize localization.
LOCALE_PATH = f'{os.path.dirname(sys.argv[0])}/locale/'
try:
    CURRENT_LANGUAGE = locale.getlocale()[0][:2]
except:
    # Fallback for old Windows versions.
    CURRENT_LANGUAGE = locale.getdefaultlocale()[0][:2]
try:
    t = gettext.translation('novelyst_updater', LOCALE_PATH, languages=[CURRENT_LANGUAGE])
    _ = t.gettext
except:

    def _(message):

        return message


class CancelCheck(Exception):
    """Exception used for cancelling the update check."""
    pass


class Plugin:
    """Template plugin class.
    
    Public methods:
        install(ui) -- Install the plugin and extend the novelyst user interface.
    """
    VERSION = '@release'
    NOVELYST_API = '4.36'
    DESCRIPTION = 'Update checker'
    URL = 'https://peter88213.github.io/novelyst_updater'
    _HELP_URL = 'https://peter88213.github.io/novelyst_updater/usage'

    def install(self, ui):
        """Install the plugin and extend the novelyst user interface.
        
        Positional arguments:
            ui -- reference to the NovelystTk instance of the application.
        """
        self._ui = ui

        # Add an entry to the Help menu.
        self._ui.helpMenu.add_command(label=_('novelyst_updater Online help'), command=lambda: webbrowser.open(self._HELP_URL))

        # Add an entry to the Tools menu.
        self._ui.toolsMenu.add_command(label=_('Check for updates'), command=self._check_for_updates)

    def _check_for_updates(self):
        """Check novelyst and all installed plugins for updates."""
        found = False
        print('Check for updates')

        # Check novelyst.
        repoName = 'novelyst'
        print(repoName)
        try:
            majorVersion, minorVersion, patchlevel, downloadUrl = self._get_version_info(repoName)
        except:
            messagebox.showerror(_('Check for updates'), _('No online update information for novelyst found.'))
            return

        try:
            latest = (majorVersion, minorVersion, patchlevel)
            print(f'Latest  : {latest}')
            current = (self._ui.plugins.majorVersion, self._ui.plugins.minorVersion, self._ui.plugins.patchlevel)
            print(f'Current : {current}')
            if self._update_available(latest, current):
                self._download_update('novelyst', downloadUrl)
                found = True

            # Check installed plugins.
            for repoName in self._ui.plugins:
                print(repoName)
                try:
                    # Latest version
                    majorVersion, minorVersion, patchlevel, downloadUrl = self._get_version_info(repoName)
                    latest = (majorVersion, minorVersion, patchlevel)
                    print(f'Latest  : {latest}')

                    # Current version
                    majorVersion, minorVersion, patchlevel = self._ui.plugins[repoName].VERSION.split('.')
                    current = (int(majorVersion), int(minorVersion), int(patchlevel))
                    print(f'Current : {current}')
                except:
                    continue
                else:
                    if self._update_available(latest, current):
                        self._download_update(repoName, downloadUrl)
                        found = True
            if not found:
                messagebox.showinfo(_('Check for updates'), _('No updates available.'))
        except CancelCheck:
            # user pressed the "cancel" button
            pass

    def _download_update(self, repo, downloadUrl):
        """Start the web browser with downloadUrl on demand.
        
        Positional arguments:
            repo: str -- Repository name of the app or plugin.
            downloadUrl: str -- Download URL of the latest release in the repository.
        
        Exceptions:
            raise CancelCheck, if the update check is to be cancelled.
        """
        text = f'{_("An update is available for")} {repo}.\n{_("Start your web browser for download?")}'
        answer = messagebox.askyesnocancel(_('Check for updates'), text)
        if answer:
            # user pressed the "Yes" button
            webbrowser.open(downloadUrl)
        elif answer is None:
            # user pressed the "Cancel" button
            raise CancelCheck

    def _get_version_info(self, repoName):
        """Return version information and download URL stored in a repository's VERSION file.
        
        Positional arguments:
            repoName: str -- The repository's name.
        
        Return major version number, minor version number, patch level, and download URL.        
        """
        versionUrl = f'https://github.com/peter88213/{repoName}/raw/main/VERSION'
        data = urlopen(versionUrl)
        versionInfo = data.read().decode('utf-8')
        config = configparser.ConfigParser()
        config.read_string(versionInfo)
        downloadUrl = config['LATEST']['download_link']
        version = config['LATEST']['version']
        majorVersion, minorVersion, patchlevel = version.split('.')
        return int(majorVersion), int(minorVersion), int(patchlevel), downloadUrl

    def _update_available(self, latest, current):
        """Return True, if the latest version number is greater than the current one."""
        if latest[0] > current[0]:
            return True

        if latest[0] == current[0]:
            if latest[1] > current[1]:
                return True

            if latest[1] == current[1]:
                if latest[2] > current[2]:
                    return True

        return False



tools/build_nv_updater.py
"""Build a nv_updater plugin.
        
In order to distribute a single script without dependencies, 
this script "inlines" all modules imported from the novxlib package.

The novxlib project (see https://github.com/peter88213/novxlib)
must be located on the same directory level as the nv_updater project. 

Copyright (c) 2024 Peter Triesberger
For further information see https://github.com/peter88213/nv_updater
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import sys
sys.path.insert(0, f'{os.getcwd()}/../../novxlib-Alpha/src')
import inliner

SRC = '../src/'
BUILD = '../test/'
SOURCE_FILE = f'{SRC}nv_updater.py'
TARGET_FILE = f'{BUILD}nv_updater.py'

os.makedirs(BUILD, exist_ok=True)


def main():
    inliner.run(SOURCE_FILE, TARGET_FILE, 'nvupdaterlib', '../../novelyst_updater/src/')
    inliner.run(TARGET_FILE, TARGET_FILE, 'noveltreelib', '../../noveltree/src/')
    inliner.run(TARGET_FILE, TARGET_FILE, 'novxlib', '../../novxlib-Alpha/src/')
    print('Done.')


if __name__ == '__main__':
    main()


tools/build_novelyst_updater.py
"""Build a novelyst_updater plugin.
        
In order to distribute a single script without dependencies, 
this script "inlines" all modules imported from the novxlib package.

The novxlib project (see https://github.com/peter88213/novxlib)
must be located on the same directory level as the novelyst_updater project. 

Copyright (c) 2023 Peter Triesberger
For further information see https://github.com/peter88213/noveltree_updater
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)
"""
import os
import sys
sys.path.insert(0, f'{os.getcwd()}/../../novxlib-Alpha/src')
import inliner

SRC = '../src/'
BUILD = '../test/'
SOURCE_FILE = f'{SRC}novelyst_updater.py'
TARGET_FILE = f'{BUILD}novelyst_updater.py'

os.makedirs(BUILD, exist_ok=True)


def main():
    inliner.run(SOURCE_FILE, TARGET_FILE, 'nvupdaterlib', '../../novelyst_updater/src/')
    inliner.run(TARGET_FILE, TARGET_FILE, 'novelystlib', '../../novelyst/src/')
    inliner.run(TARGET_FILE, TARGET_FILE, 'novxlib', '../../novxlib-Alpha/src/')
    print('Done.')


if __name__ == '__main__':
    main()


src/nv_updater.py
"""An update checker plugin for noveltree.

Requires Python 3.6+
Copyright (c) 2024 Peter Triesberger
For further information see https://github.com/peter88213/nv_updater
License: GNU GPLv3 (https://www.gnu.org/licenses/gpl-3.0.en.html)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

"""
import sys
import os
from tkinter import messagebox
import locale
import gettext
from urllib.request import urlopen
import webbrowser
import configparser

# Initialize localization.
LOCALE_PATH = f'{os.path.dirname(sys.argv[0])}/locale/'
try:
    CURRENT_LANGUAGE = locale.getlocale()[0][:2]
except:
    # Fallback for old Windows versions.
    CURRENT_LANGUAGE = locale.getdefaultlocale()[0][:2]
try:
    t = gettext.translation('nv_updater', LOCALE_PATH, languages=[CURRENT_LANGUAGE])
    _ = t.gettext
except:

    def _(message):

        return message


class CancelCheck(Exception):
    """Exception used for cancelling the update check."""
    pass


class Plugin:
    """Template plugin class.
    
    Public methods:
        install(ui) -- Install the plugin and extend the noveltree user interface.
    """
    VERSION = '@release'
    NOVELYST_API = '0.1'
    DESCRIPTION = 'Update checker'
    URL = 'https://peter88213.github.io/nv_updater'
    _HELP_URL = 'https://peter88213.github.io/nv_updater/usage'

    def install(self, ui):
        """Install the plugin and extend the noveltree user interface.
        
        Positional arguments:
            ui -- reference to the NoveltreeUi instance of the application.
        """
        self._ui = ui

        # Add an entry to the Help menu.
        self._ui.helpMenu.add_command(label=_('nv_updater Online help'), command=lambda: webbrowser.open(self._HELP_URL))

        # Add an entry to the Tools menu.
        self._ui.toolsMenu.add_command(label=_('Check for updates'), command=self._check_for_updates)

    def _check_for_updates(self):
        """Check noveltree and all installed plugins for updates."""
        found = False
        print('Check for updates')

        # Check noveltree.
        repoName = 'noveltree'
        print(repoName)
        try:
            majorVersion, minorVersion, patchlevel, downloadUrl = self._get_version_info(repoName)
        except:
            messagebox.showerror(_('Check for updates'), _('No online update information for noveltree found.'))
            return

        try:
            latest = (majorVersion, minorVersion, patchlevel)
            print(f'Latest  : {latest}')
            current = (self._ui.plugins.majorVersion, self._ui.plugins.minorVersion, self._ui.plugins.patchlevel)
            print(f'Current : {current}')
            if self._update_available(latest, current):
                self._download_update('noveltree', downloadUrl)
                found = True

            # Check installed plugins.
            for repoName in self._ui.plugins:
                print(repoName)
                try:
                    # Latest version
                    majorVersion, minorVersion, patchlevel, downloadUrl = self._get_version_info(repoName)
                    latest = (majorVersion, minorVersion, patchlevel)
                    print(f'Latest  : {latest}')

                    # Current version
                    majorVersion, minorVersion, patchlevel = self._ui.plugins[repoName].VERSION.split('.')
                    current = (int(majorVersion), int(minorVersion), int(patchlevel))
                    print(f'Current : {current}')
                except:
                    continue
                else:
                    if self._update_available(latest, current):
                        self._download_update(repoName, downloadUrl)
                        found = True
            if not found:
                messagebox.showinfo(_('Check for updates'), _('No updates available.'))
        except CancelCheck:
            # user pressed the "cancel" button
            pass

    def _download_update(self, repo, downloadUrl):
        """Start the web browser with downloadUrl on demand.
        
        Positional arguments:
            repo: str -- Repository name of the app or plugin.
            downloadUrl: str -- Download URL of the latest release in the repository.
        
        Exceptions:
            raise CancelCheck, if the update check is to be cancelled.
        """
        text = f'{_("An update is available for")} {repo}.\n{_("Start your web browser for download?")}'
        answer = messagebox.askyesnocancel(_('Check for updates'), text)
        if answer:
            # user pressed the "Yes" button
            webbrowser.open(downloadUrl)
        elif answer is None:
            # user pressed the "Cancel" button
            raise CancelCheck

    def _get_version_info(self, repoName):
        """Return version information and download URL stored in a repository's VERSION file.
        
        Positional arguments:
            repoName: str -- The repository's name.
        
        Return major version number, minor version number, patch level, and download URL.        
        """
        versionUrl = f'https://github.com/peter88213/{repoName}/raw/main/VERSION'
        data = urlopen(versionUrl)
        versionInfo = data.read().decode('utf-8')
        config = configparser.ConfigParser()
        config.read_string(versionInfo)
        downloadUrl = config['LATEST']['download_link']
        version = config['LATEST']['version']
        majorVersion, minorVersion, patchlevel = version.split('.')
        return int(majorVersion), int(minorVersion), int(patchlevel), downloadUrl

    def _update_available(self, latest, current):
        """Return True, if the latest version number is greater than the current one."""
        if latest[0] > current[0]:
            return True

        if latest[0] == current[0]:
            if latest[1] > current[1]:
                return True

            if latest[1] == current[1]:
                if latest[2] > current[2]:
                    return True

        return False

